### ISSUE TITLE
Text Position and Scaling Issues

### ISSUE THREAD

**Issue by SolarLune** [GoranM/bdx#830]
Hey, there. So I'm posting about the position and scaling issues with the Text Object. Here's a screenshot to show the first problems I've noticed.

![screenshot from 2016-11-14 21-20-31](https://cloud.githubusercontent.com/assets/4733521/20294060/45895aca-aab0-11e6-97b5-4d718e103938.png)

On the left, the scene in Blender can be seen. The text is approximately 2 Blender Units in length. On the right, the scene in BDX can be seen. The text is incorrectly sized, and does not "start" on the base line of the text line.

To resolve these two issues, what should be done? Simply sliding the text mesh vertices upwards to sit on the line rather than sit in the middle of it might be a good idea to start with.

**Comment by GoranM** [comment]
As stated in [the wiki](https://github.com/GoranM/bdx/wiki/General-Overview#text):

> To compensate for those discrepancies, you can adjust the font size and paragraph offset values in text properties, until what you see in blender is close enough to what you see in BDX.

**Comment by SolarLune** [comment]
I know that, but that's the problem. This is also coupled with the fact that larger font size exports change the resolution of the text, but also the size of the text object when rendered in BDX.

Is it simply impossible to get the shape and scale of text in Blender to be the same for BDX?

**Comment by Lucrecious** [comment]
Can't you do it manually? Get the size of the text in Blender and put it in the json and adjust realtime? This is assuming that you can get text width and height while in Libgdx so you can manually change the size. If I recall, the reason we couldn't do it before was because there was no access to text dimensions in real-time.

**Comment by Lucrecious** [comment]
You can probably get the text dimensions by getting the distance between the "leftmost" vertex and the "rightmost", and the "upmost" and "downmost". Then adjust to what the json says the dimensions are supposed to be

**Comment by GoranM** [comment]
> Is it simply impossible to get the shape and scale of text in Blender to be the same for BDX?

It's not impossible, but you need to know precisely how Blender interprets ttf font data when rendering font objects, and make that match with how BDX font objects are formed.

I was hoping that there was just some constant multiplier that could be found and applied to bridge the gap, but, in my experiments, the values seem to differ somewhat depending on the font, so it appears that it's not that simple.

**Comment by SolarLune** [comment]
Hey, so I made progress on this - with this commit, the difference between Blender and in-game text objects is very minimal. Lemme know your thoughts.

https://github.com/SolarLune/bdx/commit/c5b380fa649f5cc9dcd0b43e0d88ad96286afb06

**Comment by SolarLune** [comment]
OK, so more progress - it's better, but still not exactly perfect. Some fonts are pretty much spot on:

![screenshot from 2017-07-04 08 33 52](https://user-images.githubusercontent.com/4733521/27836283-9ddd3b4e-6093-11e7-9d45-52c284cbe504.png)

while others (like BFont) are a bit more spotty:

![screenshot from 2017-07-04 08 33 37](https://user-images.githubusercontent.com/4733521/27836288-a270272a-6093-11e7-8e36-c2f7c8d4544d.png)

I think it might have to do with the way Blender displays line heights - I think it's different from the values LibGDX records in its font files.

**Comment by rafcolson** [comment]
Yeah, I remember I had the same results when reviewing this. I used a constant value for the blender font to fix the problem. But it somehow come out a little different than the constant @GoranM was using. It was not `0.0225f * 1.4f` or `0.0315` but `0.03125` or `1/32` which makes sense since the scale should be pixel based.

**Comment by GoranM** [comment]
If Blender incorporates current resolution information when rendering fonts, it's going to differ across machines.

So, someone could get something that looks more accurate on their machine, but there's no guarantee that it will look more accurate on another.

**Comment by SolarLune** [comment]
Do you mean screen resolution? If so, I'll try testing at different resolutions to see if the fonts appear to stay the same.

**Comment by SolarLune** [comment]
Yo, I forgot about this. The screen resolution changes didn't do anything to change how the fonts render. The result I last had was close enough to commit, I think. I'll submit a PR for it.
